src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -448,11 +453,42 @@ def _do_upgrade(self):
             ):
                 return
 
+            to_upgrade = []
+            known_ok_to_stop: List[str] = []
             for d in need_upgrade:
                 assert d.daemon_type is not None
                 assert d.daemon_id is not None
                 assert d.hostname is not None
 
+                if not d.container_image_id:
+                    if d.container_image_name == target_image:
+                        logger.debug(
+                            'daemon %s has unknown container_image_id but has correct image name' % (d.name()))
+                        continue
+
+                if known_ok_to_stop:
+                    if d.name() in known_ok_to_stop:
+                        logger.info(f'Upgrade: {d.name()} is also safe to restart')
+                        to_upgrade.append(d)
+                    continue
+
+                if not self._wait_for_ok_to_stop(d, known_ok_to_stop):
```

Comment:
took me a while to figure out that _wait_for_ok_to_stop changes known_ok_to_stop in this loop

Reply:
added some comments!